test/e2e/network/service.go

```diff
@@ -1883,3 +1899,81 @@ func execSourceipTest(f *framework.Framework, c clientset.Interface, ns, nodeNam
 	}
 	return execPod.Status.PodIP, outputs[1]
 }
+
+func execAffinityTestForNonLBService(f *framework.Framework, cs clientset.Interface, serviceType v1.ServiceType) {
+	ns := f.Namespace.Name
+	numPods, servicePort := 3, 80
+	serviceName := "service"
+	By("creating service in namespace " + ns)
+	_, _, err := framework.StartCustomizedServeHostnameService(cs, f.InternalClientset, ns, serviceName, servicePort, numPods, func(svc *v1.Service) {
+		svc.Spec.Type = serviceType
+	})
+	Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		framework.StopServeHostnameService(cs, f.InternalClientset, ns, serviceName)
+	}()
+	jig := framework.NewServiceTestJig(cs, serviceName)
+	svc, err := jig.Client.CoreV1().Services(ns).Get(serviceName, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	var svcIpPort string
+	if serviceType == v1.ServiceTypeNodePort {
+		nodeIP := framework.PickNodeIP(jig.Client)
+		svcIpPort = net.JoinHostPort(nodeIP, strconv.Itoa(int(svc.Spec.Ports[0].NodePort)))
+	} else {
+		svcIpPort = net.JoinHostPort(svc.Spec.ClusterIP, strconv.Itoa(servicePort))
+	}
+
+	execPodName := framework.CreateExecPodOrFail(cs, ns, "execpod-affinity", nil)
+	defer func() {
+		framework.Logf("Cleaning up the exec pod")
+		err := cs.CoreV1().Pods(ns).Delete(execPodName, nil)
+		Expect(err).NotTo(HaveOccurred())
+	}()
+	execPod, err := cs.CoreV1().Pods(ns).Get(execPodName, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	Expect(framework.TestAffinityFromPod(execPod, svcIpPort, false)).To(BeTrue(), "Affinity shouldn't hold or connection timeout.")
+	svc = jig.UpdateServiceOrFail(svc.Namespace, svc.Name, func(svc *v1.Service) {
+		svc.Spec.SessionAffinity = v1.ServiceAffinityClientIP
+	})
+	Expect(framework.TestAffinityFromPod(execPod, svcIpPort, true)).To(BeTrue(), "Affinity didn't hold or connection timeout.")
+}
+
+func execAffinityTestForLBService(f *framework.Framework, cs clientset.Interface, externalTrafficPolicy v1.ServiceExternalTrafficPolicyType) {
+	ns := f.Namespace.Name
+	numPods, servicePort := 3, 80
+	serviceName := "service"
+
+	By("creating service in namespace " + ns)
+	_, _, err := framework.StartCustomizedServeHostnameService(cs, f.InternalClientset, ns, serviceName, servicePort, numPods, func(svc *v1.Service) {
+		svc.Spec.Type = v1.ServiceTypeLoadBalancer
+		svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeLocal
+	})
+	Expect(err).NotTo(HaveOccurred())
+	jig := framework.NewServiceTestJig(cs, serviceName)
+	By("waiting for loadbalancer for service " + ns + "/" + serviceName)
+	svc := jig.WaitForLoadBalancerOrFail(ns, serviceName, framework.LoadBalancerCreateTimeoutDefault)
+	jig.SanityCheckService(svc, v1.ServiceTypeLoadBalancer)
+	defer func() {
+		jig.ChangeServiceType(svc.Namespace, svc.Name, v1.ServiceTypeClusterIP, framework.LoadBalancerCreateTimeoutDefault)
+		Expect(cs.CoreV1().Services(svc.Namespace).Delete(svc.Name, nil)).NotTo(HaveOccurred())
```

Comment:
Minor: Sorry I missed the point of not defer `StopServeHostnameService` instead...`StartCustomizedServeHostnameService` also creates replication controller and that needs to be cleaned up.

Reply:
Oh, there's RC left. 

If StopServeHostnameService releases the loadbalancer, then we are safe to use StopServeHostnameService to release the service with RC.

Reply:
Yeah I think `StopServeHostnameService` would be enough (it deletes the service as well). The `jig.ChangeServiceType` line exists for some legacy reasons, don't think it is required anymore.